// @celo/utils/lib/attestations.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var Web3Utils = __importStar(require("web3-utils"));
var address_1 = require("./address");
var signatureUtils_1 = require("./signatureUtils");
var IdentifierType;
(function (IdentifierType) {
    IdentifierType[IdentifierType["PHONE_NUMBER"] = 0] = "PHONE_NUMBER";
})(IdentifierType || (IdentifierType = {}));
function hashIdentifier(identifier, type) {
    switch (type) {
        case IdentifierType.PHONE_NUMBER:
            return Web3Utils.soliditySha3({ type: 'string', value: identifier });
        default:
            return '';
    }
}
function getAttestationMessageToSignFromIdentifier(identifier, account) {
    return getAttestationMessageToSignFromPhoneHash(hashIdentifier(identifier, IdentifierType.PHONE_NUMBER), account);
}
exports.getAttestationMessageToSignFromIdentifier = getAttestationMessageToSignFromIdentifier;
function getAttestationMessageToSignFromPhoneHash(phoneHash, account) {
    var messageHash = Web3Utils.soliditySha3({ type: 'bytes32', value: phoneHash }, { type: 'address', value: account });
    return messageHash;
}
exports.getAttestationMessageToSignFromPhoneHash = getAttestationMessageToSignFromPhoneHash;
function base64ToHex(base64String) {
    return '0x' + Buffer.from(base64String, 'base64').toString('hex');
}
exports.base64ToHex = base64ToHex;
function attestToIdentifier(identifier, account, privateKey) {
    var issuer = address_1.privateKeyToAddress(privateKey);
    var _a = signatureUtils_1.SignatureUtils.signMessage(getAttestationMessageToSignFromIdentifier(identifier, account), privateKey, issuer), v = _a.v, r = _a.r, s = _a.s;
    return { v: v, r: r, s: s };
}
exports.attestToIdentifier = attestToIdentifier;
function sanitizeMessageBase64(base64String) {
    // Replace occurrences of ¿ with _. Unsure why that is happening right now
    return base64String.replace(/(¿|§)/gi, '_');
}
exports.sanitizeMessageBase64 = sanitizeMessageBase64;
var attestationCodeRegex = new RegExp(/(.* |^)(?:celo:\/\/wallet\/v\/)?([a-zA-Z0-9=\+\/_-]{87,88})($| .*)/);
function messageContainsAttestationCode(message) {
    return attestationCodeRegex.test(message);
}
exports.messageContainsAttestationCode = messageContainsAttestationCode;
function extractAttestationCodeFromMessage(message) {
    var sanitizedMessage = sanitizeMessageBase64(message);
    if (!messageContainsAttestationCode(sanitizedMessage)) {
        return null;
    }
    var matches = sanitizedMessage.match(attestationCodeRegex);
    if (!matches || matches.length < 3) {
        return null;
    }
    return base64ToHex(matches[2]);
}
exports.extractAttestationCodeFromMessage = extractAttestationCodeFromMessage;
exports.AttestationUtils = {
    getAttestationMessageToSignFromIdentifier: getAttestationMessageToSignFromIdentifier,
    getAttestationMessageToSignFromPhoneHash: getAttestationMessageToSignFromPhoneHash,
    base64ToHex: base64ToHex,
    attestToIdentifier: attestToIdentifier,
    sanitizeMessageBase64: sanitizeMessageBase64,
    messageContainsAttestationCode: messageContainsAttestationCode,
    extractAttestationCodeFromMessage: extractAttestationCodeFromMessage,
};
